[docs-only][ADR] Index and store metadata (#7515)
owncloud/ocis: docs/ocis/adr/0023-index-and-store-metadata.md

like artist, album and title or exif metadata in images.

## Decision Drivers <!-- optional -->

## Considered Options

* [Store subset of extracted metadata required for graph api](#store-subset-of-extracted-metadata-required-for-graph-api)
* [Store subset of extracted metadata specified by another standard](#store-subset-of-extracted-metadata-specified-by-another-standard)
* [Store everything from extractors](#store-everything-from-extractors)

## Decision Outcome

Chosen option: "[store only subset of extracted metadata required for graph api](#store-subset-of-extracted-metadata-required-for-graph-api)", because Graph API is a simple common denominator and we want to avoid putting the complexity of mapping non-standardized data from potentially different extractors in several areas of the code base. Storage and index keys are determined by facet and property name, e.g. `audio.artist` for the artist in a music file. Storage keys are additionally prefixed with `libre.graph.`, i.e. `libre.graph.audio.artist`.

### Positive Consequences:

* Graph API endpoint implementation is trivial

> * documented public api and stored data are the same

Reviewer: ```suggestion
* Documented public api and stored data are the same
```

* reasonable complexity for the initial implementation

### Negative consequences:

* Graph API is limited, so not *all* available metadata can be accessed

> * switching the internal format and adding more metadata later will require re-indexing

Reviewer: ```suggestion
* Switching the internal format and adding more metadata later will require re-indexing
```

---
title: "23. Index and store metadata"
date: 2023-10-17T15:15:00+01:00
weight: 23
geekdocRepo: https://github.com/owncloud/ocis
geekdocEditPath: edit/master/docs/ocis/adr
geekdocFilePath: 0023-index-and-store-metadata.md
---

* Status: proposed
* Deciders: @butonic, @micbar
* Date: 2023-10-17

## Context and Problem Statement

ownCloud Infinite Scale is supposed to become a data platform and as such it needs to provide access to metadata.
Currently only metadata available for all file types (filesize, mime-type, ...) is stored in the index and the metadata storage.
We want to make other metadata available to consumers of our internal and external APIs. Simple examples would be audio metadata

> like artist, album and title or exif metadata in images.

Reviewer 2: This I find misleading. In our discussion you said that you were focusing on a set of metadata for known file types (audio...) that have a representation in the standard Graph API as of today.

This does not talk about a "generic" interface for meta data. I think that would be important to mention.

Author: fair point, I'll try to improve

Author: I've read it a few times again and I don't think it's misleading. That we focus on the graph api is not part of the problem statement, but part of the decision imho... what do you think?

FWIW I've slightly adjusted the wording here and added a note to the outcome.

---
title: "23. Index and store metadata"
date: 2023-10-17T15:15:00+01:00
weight: 23
geekdocRepo: https://github.com/owncloud/ocis
geekdocEditPath: edit/master/docs/ocis/adr
geekdocFilePath: 0023-index-and-store-metadata.md
---

* Status: proposed
* Deciders: @butonic, @micbar
* Date: 2023-10-17

## Context and Problem Statement

ownCloud Infinite Scale is supposed to become a data platform and as such it needs to provide access to metadata.
Currently only metadata common to all file types (filesize, mime-type, ...) is stored in the index and the metadata storage.

> We want to make other file type specific metadata available to consumers of our internal and external APIs.
> Simple examples would be audio metadata like artist, album and title or exif metadata in images.

Reviewer 2: ```suggestion
Metadata of audio- and image data should be made available to consumers of our internal and external APIs.
For both types the Graph API 1.0 has pre defined facets.

This ADR is not talking about a generic metadata storing and retrieval system.
```

Author: I think I still disagree, this is putting too much assumptions into the context and problem statement - the context is that we want to make metadata available, the outcome is that we want to focus on graph api because it's simple and reasonable to implement.

This ADR *is* talking about generic metadata storage/retrieval, it explains why we decided against that for now

Author: see `### Store everything from extractors`

Reviewer: i agree with you @dschmidt

* [Store subset of extracted metadata specified by another standard](#store-subset-of-extracted-metadata-specified-by-another-standard)
* [Store everything from extractors](#store-everything-from-extractors)

## Decision Outcome

Chosen option: "[store only subset of extracted metadata required for graph api](#store-subset-of-extracted-metadata-required-for-graph-api)", because Graph API is a simple common denominator and we want to avoid putting the complexity of mapping non-standardized data from potentially different extractors in several areas of the code base. Storage and index keys are determined by facet and property name, e.g. `audio.artist` for the artist in a music file. Storage keys are additionally prefixed with `libre.graph.`, i.e. `libre.graph.audio.artist`.
Handling Graph API specific metadata is a first step towards handling metadata. More generic and extensible handling of arbitrary metadata can be added later.

### Positive Consequences:

* Graph API endpoint implementation is trivial
* Documented public api and stored data are the same
* Reasonable complexity for the initial implementation

> ### Negative consequences:

Reviewer 3: ```suggestion
### Negative Consequences:
```

* Graph API is limited, so not *all* available metadata can be accessed
* Switching the internal format and adding more metadata later will require re-indexing

## Pros and Cons of the Options <!-- optional -->

> ### Store subset of extracted metadata required for Graph API

Reviewer 3: ```suggestion
### Store Subset of Extracted Metadata Required for the Graph API
```

Use Graph API facets and properties for determining the subset of stored metadata and the storage key.
The index key for the `artist` property of the `audio` facet is `audio.artist`, the storage key is additionally prefixed with `libre.graph.`.

* Good, because central mapping of values happens consistently and only once in a central place
    - it happens in the extractor (integration) which likely knows best how to map metadata to standard properties
* Good, because when multiple extractors share a common set of provided values, applications can rely on the mapping and the complexity is kept low
* Bad, because not all metadata is available, not everything can be searched
* Good, because Graph API already chose a reasonable subset of most interesting properties

### Store subset of extracted metadata specified by another standard

> There are a bunch of metadata standards but none of them is really universal, there is always something that is only supported in or the other standard.
> Tika for example extracts audio metadata using a mixture of Dublin Core and XMP Dynamic Media keys.

Reviewer 3: ```suggestion
There are a bunch of metadata standards but none of them is really universal. There is always something that is only supported in or the other standard. Tika for example extracts audio metadata using a mixture of Dublin Core and XMP Dynamic Media keys.
```

- Bad, because it makes implementing a new extractor integration harder
- Bad, because it makes using the stored data more complicated than a simple standard like discussed above

> ### Store everything from extractors

Reviewer 3: ```suggestion
### Store Everything from Extractors
```
